server/lib/src/valueset/image.rs

```diff
@@ -0,0 +1,663 @@
+        let mut set = HashSet::new();
+        set.insert(image);
+        Box::new(ValueSetImage { set })
+    }
+
+    // pushing
+    pub fn push(&mut self, image: ImageValue) -> bool {
+        match image.validate_image() {
+            Ok(_) => self.set.insert(image),
+            Err(err) => {
+                admin_error!(
+                    "Image didn't pass validation, not adding to value! Error: {}",
+                    err
+                );
+                false
+            }
+        }
+    }
+
+    pub fn from_dbvs2(data: &[DbValueImage]) -> Result<ValueSet, OperationError> {
+        Ok(Box::new(ValueSetImage {
+            set: data
+                .iter()
+                .cloned()
+                .map(|e| match e {
+                    DbValueImage::V1 {
+                        filename,
+                        filetype,
+                        contents,
+                    } => ImageValue::new(filename, filetype, contents),
+                })
+                .collect(),
+        }))
+    }
+
+    pub fn from_repl_v1(data: &[DbValueImage]) -> Result<ValueSet, OperationError> {
+        let mut set: HashSet<ImageValue> = HashSet::new();
+        for image in data {
+            let image = match image.clone() {
+                DbValueImage::V1 {
+                    filename,
+                    filetype,
+                    contents,
+                } => ImageValue::new(filename, filetype, contents),
+            };
+            if image.validate_image().is_ok() {
+                set.insert(image.clone());
+            } else {
+                admin_error!("Image didn't pass validation, not adding to value!");
+                return Err(OperationError::InvalidValueState);
+            }
+        }
+
+        Ok(Box::new(ValueSetImage { set }))
+    }
+
+    // We need to allow this, because rust doesn't allow us to impl FromIterator on foreign
+    // types, and `ImageValue` is foreign.
+    #[allow(clippy::should_implement_trait)]
+    pub fn from_iter<T>(iter: T) -> Option<Box<ValueSetImage>>
+    where
+        T: IntoIterator<Item = ImageValue>,
+    {
+        let mut set: HashSet<ImageValue> = HashSet::new();
+        for image in iter {
+            match image.validate_image() {
+                Ok(_) => set.insert(image),
+                Err(err) => {
+                    admin_error!(
+                        "Image didn't pass validation, not adding to value! Error: {}",
+                        err
+                    );
+                    return None;
+                }
+            };
+        }
+        Some(Box::new(ValueSetImage { set }))
+    }
+}
+
+impl ValueSetT for ValueSetImage {
+    fn insert_checked(&mut self, value: Value) -> Result<bool, OperationError> {
+        match value {
+            Value::Image(image) => match image.validate_image() {
+                Ok(_) => Ok(self.set.insert(image)),
+                Err(err) => {
+                    admin_error!(
```

Comment:
There is a separate validation phase though, so is this needed here?

Reply:
I'm pretty sure this could just call self.set.clear - it's no-mv, so if you want to remove anything then it's only going to be one thing

Reply:
Well, we need to assume for now it could be mv. I want to change syntax of attrs in future that the attr type defines if it's mv or not, but for now we're stuck wuth everything being possible mv.

Reply:
ezpz, can sort that out